src/oph/ehoks/palaute/opiskelija.clj

```diff
@@ -193,29 +193,38 @@
               fully migrated to eHOKS."
   ([kysely hoks] (initiate-if-needed! kysely hoks nil))
   ([kysely hoks opts]
-    (jdbc/with-db-transaction
-      [tx db/spec]
-      (let [opiskeluoikeus
-            (koski/get-existing-opiskeluoikeus! (:opiskeluoikeus-oid hoks))
-            koulutustoimija
-            (palaute/koulutustoimija-oid! opiskeluoikeus)
-            existing-heratteet
-            (existing-heratteet! kysely hoks koulutustoimija tx)
-            [init-state field reason]
-            (initial-palaute-state-and-reason
-              kysely hoks opiskeluoikeus
-              (when-not (:resend? opts) existing-heratteet))]
-        (log/info "Initial state for" kysely "for HOKS" (:id hoks)
-                  "will be" (or init-state :ei-luoda-ollenkaan)
-                  "because of" reason "in" field)
-        (when init-state
-          (initiate! kysely hoks opiskeluoikeus koulutustoimija tx
-                     (assoc opts
-                            :initial-state init-state
-                            :reason reason
-                            :other-info (select-keys hoks [field])
-                            :existing-heratteet existing-heratteet)))
-        init-state))))
+    (try
+      (jdbc/with-db-transaction
+        [tx db/spec]
+        (let [opiskeluoikeus
+              (koski/get-existing-opiskeluoikeus! (:opiskeluoikeus-oid hoks))
+              koulutustoimija
+              (palaute/koulutustoimija-oid! opiskeluoikeus)
+              existing-heratteet
+              (existing-heratteet! kysely hoks koulutustoimija tx)
+              [init-state field reason]
+              (initial-palaute-state-and-reason
+                kysely hoks opiskeluoikeus
+                (when-not (:resend? opts) existing-heratteet))]
+          (log/info "Initial state for" kysely "for HOKS" (:id hoks)
+                    "will be" (or init-state :ei-luoda-ollenkaan)
+                    "because of" reason "in" field)
+          (when init-state
+            (initiate! kysely hoks opiskeluoikeus koulutustoimija tx
+                       (assoc opts
+                              :initial-state init-state
+                              :reason reason
+                              :other-info (select-keys hoks [field])
+                              :existing-heratteet existing-heratteet)))
+          init-state))
+      (catch clojure.lang.ExceptionInfo e
+        (if (= ::koski/opiskeluoikeus-not-found (:type (ex-data e)))
+          ;; this should only happen when initiate-if-needed! is called in
+          ;; batch; for HOKS saving, a missing opiskeluoikeus prevents it
+          ;; already earlier
+          (log/warn "Opiskeluoikeus" (:opiskeluoikeus-oid hoks) "not found"
+                    "for hoks" (:id hoks) "so not sending" kysely "for it")
+          (throw e))))))
```

Comment:
Idea: Poikkeuksen kiinninappaaminen on tässä toimiva, mutta oisko parempi käyttää `koski/get-opiskeluoikeus!`?
```suggestion
    (jdbc/with-db-transaction
      [tx db/spec]
      (if-let [opiskeluoikeus (koski/get-opiskeluoikeus!
                                (:opiskeluoikeus-oid hoks))]
        (let [koulutustoimija    (palaute/koulutustoimija-oid! opiskeluoikeus)
              existing-heratteet (existing-heratteet!
                                  kysely hoks koulutustoimija tx)
              [init-state field reason]
              (initial-palaute-state-and-reason
                kysely hoks opiskeluoikeus
                (when-not (:resend? opts) existing-heratteet))]
          (log/info "Initial state for" kysely "for HOKS" (:id hoks)
                    "will be" (or init-state :ei-luoda-ollenkaan)
                    "because of" reason "in" field)
          (when init-state
            (initiate! kysely hoks opiskeluoikeus koulutustoimija tx
                        (assoc opts
                              :initial-state init-state
                              :reason reason
                              :other-info (select-keys hoks [field])
                              :existing-heratteet existing-heratteet)))
          init-state)
        ;; this should only happen when initiate-if-needed! is called in
        ;; batch; for HOKS saving, a missing opiskeluoikeus prevents it
        ;; already earlier
        (log/warn "Opiskeluoikeus" (:opiskeluoikeus-oid hoks) "not found"
                  "for hoks" (:id hoks) "so not sending" kysely "for it")))
```

Reply:
Tässä tapauksessa poikkeukseen perustuva control flow on mun mielestä suunnilleen yhtä hyvä kuin normaali, koska kyse on tosiaan aika poikkeuksellisesta tilanteesta (Pythonin käytäntö: laita "tavallisin polku" eli happy case `try`-blokkiin, ja anna poikkeusten tapahtua.  Tee vähemmän tavalliset polut poikkeuskäsittelijöissä, niin ne ei hämärrä perustapausta.)  Mutta ei oo mitään erityistä syytä myöskään suosia poikkeusten käyttöä.

`if-let` ei miellytä mua tässä, koska sen eka lauseke on sairaan pitkä mikä tarkoittaa että tokasta lausekkeesta (tuosta lokituksesta) on raskasta lukea, minkä iffin else-haara tää nyt onkaan.

Mut toisaalta mietin, että puuttuva opiskeluoikeus on yks syy olla tekemättä herätettä joten sen vois vaan sisällyttää initial-state-and-reason:iin.  Taidan kokeilla sitä, niin voit sanoa, pidätkö tuloksena olevasta koodista.

Reply:
Onks tää uusi versio hyvä?